### PR TITLE
[sw/ottf] implemented default ISRs

### PR DESCRIPTION
**Pull request by timothytrippel**
This adds the following to the OTTF:

- defualt (weak) ISRs
- FreeRTOS context saving (asm) subroutines to be called on ISR entry/exit

Additionally, this depends on #7938, thus has duplicate commits which will be removed when merged.

### CODE REVIEW

File: sw/device/lib/testing/test_framework/ottf_isrs.c [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/test_framework/FreeRTOSConfig.h"
+
+static uint32_t get_mtval(void) {
+  uint32_t mtval;
+  CSR_READ(CSR_REG_MTVAL, &mtval);
+  return mtval;
+}
+
+__attribute__((weak)) void ottf_exception_handler(void) {
```
Comment (mcy) [inline sw/device/lib/testing/test_framework/ottf_isrs.c:17]: Nit: Can we lift `((weak))` into `OT_ATTR_WEAK` or something like that in macros.h?

File: sw/device/lib/testing/test_framework/ottf_isrs.c [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  uint32_t mcause;
+  ottf_exc_id_t exc_cause;
+
+  CSR_READ(CSR_REG_MCAUSE, &mcause);
+  exc_cause = (ottf_exc_id_t)(mcause & kIdMax);
+
+  switch (exc_cause) {
```
Comment (mcy) [inline sw/device/lib/testing/test_framework/ottf_isrs.c:24]: This is only used once. Inline it?

File: sw/device/lib/testing/test_framework/ottf_isrs.c [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+      break;
+    case kMachineECall:
+      ottf_machine_ecall();
+      break;
+    case kUserECall:
+      ottf_user_ecall();
+      break;
+    default:
+      LOG_INFO("Unknown exception triggered!");
+      while (1) {
```
Comment (mcy) [inline sw/device/lib/testing/test_framework/ottf_isrs.c:51]: `s/1/true`. The trailing semi is superfluous.

Also, make sure to include a `wait_for_interrupt()` in the loop body; infinite empty loops are UB and Clang will delete it.

Applies throughout. (Maybe worth defining a `static noreturn void loop_forever();`...)

File: sw/device/lib/testing/test_framework/ottf.h [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+ * Entry point for a SW on-device (or chip-level) test.
+ *
+ * This function should be defined externally in a standalone SW test, linked
+ * together with this library. This library provides a `main()` function that
+ * does test harness setup, initializes FreeRTOS, and starts a FreeRTOS task
+ * that executes `test_main()`.
+ *
+ * @return success or failure of the test as boolean.
+ */
+// extern bool test_main(void);
```
Comment (mcy) [inline sw/device/lib/testing/test_framework/ottf.h:66]: Hmm, did you mean to delete this?

File: sw/device/lib/testing/test_framework/freertos_port.c [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+  // Unreachable.
+  return pdFAIL;
+}
+
+void vPortEndScheduler(void) {
+  // Not implemented.
+  // TODO: trigger this to be called when from the idle task hook when all tests
+  // have completed.
+  for (;;)
```
Comment (mcy) [inline sw/device/lib/testing/test_framework/freertos_port.c:101]: Here too, make sure to do _something_ in the loop body.

File: sw/device/lib/testing/test_framework/freertos_port.c [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  CHECK(dif_rv_timer_irq_enable(&timer, kTimerHartId, kTimerComparatorId,
+                                kDifRvTimerEnabled) == kDifRvTimerOk);
+  CHECK(dif_rv_timer_arm(&timer, kTimerHartId, kTimerComparatorId,
+                         kTimerDeadline) == kDifRvTimerOk);
+
+  CHECK(dif_rv_timer_counter_set_enabled(&timer, kTimerHartId,
+                                         kDifRvTimerEnabled) == kDifRvTimerOk);
+}
+
+#endif
```
Comment (mcy) [inline sw/device/lib/testing/test_framework/freertos_port.c:77]: Make sure to mark the `endif` with the condition.

File: sw/device/lib/testing/test_framework/freertos_port.c [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"  // Generated.
+
+// NOTE: some of the function names below do NOT, and cannot, conform to the
+// style guide, since they are specific implementations of FreeRTOS defined
+// functions.
+
+// ----------------------------------------------------------------------------
+// Timer Setup (for use when preemptive scheduling is enabled)
+// ----------------------------------------------------------------------------
+#if defined(configUSE_PREEMPTION) && configUSE_PREEMPTION == 1
```
Comment (mcy) [inline sw/device/lib/testing/test_framework/freertos_port.c:24]: Is `#ifdef <macro>` not enough?

**Review by mcy — APPROVED** [review]
Looks great!

File: sw/device/lib/testing/test_framework/freertos_port.c (hunk 7 above)
Comment (timothytrippel) [reply sw/device/lib/testing/test_framework/freertos_port.c]: After a quick check ([sw/vendor/freertos_freertos_kernel/include/FreeRTOS.h:95](https://github.com/lowRISC/opentitan/blob/master/sw/vendor/freertos_freertos_kernel/include/FreeRTOS.h#L95)) FreeRTOS _requires_ configUSE_PREEMPTION be defined, so I simplified this to "#if configUSE_PREEMPTION == 1"

File: sw/device/lib/testing/test_framework/freertos_port.c (hunk 5 above)
Comment (timothytrippel) [reply sw/device/lib/testing/test_framework/freertos_port.c]: Fixed.

File: sw/device/lib/testing/test_framework/ottf.h (hunk 4 above)
Comment (timothytrippel) [reply sw/device/lib/testing/test_framework/ottf.h]: Fixed.

File: sw/device/lib/testing/test_framework/ottf_isrs.c (hunk 3 above)
Comment (timothytrippel) [reply sw/device/lib/testing/test_framework/ottf_isrs.c]: Fixed.

File: sw/device/lib/testing/test_framework/ottf_isrs.c (hunk 2 above)
Comment (timothytrippel) [reply sw/device/lib/testing/test_framework/ottf_isrs.c]: Fixed.

File: sw/device/lib/testing/test_framework/ottf_isrs.c (hunk 1 above)
Comment (timothytrippel) [reply sw/device/lib/testing/test_framework/ottf_isrs.c]: Fixed.

File: sw/device/lib/testing/test_framework/ottf_isrs.c [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/test_framework/FreeRTOSConfig.h"
+
+static uint32_t get_mtval(void) {
+  uint32_t mtval;
+  CSR_READ(CSR_REG_MTVAL, &mtval);
+  return mtval;
+}
+
+static noreturn void loop_forever(void) {
```
Comment (sriyerg) [inline sw/device/lib/testing/test_framework/ottf_isrs.c:19]: Doesn't `abort()` already do this?

File: sw/device/lib/testing/test_framework/ottf_isrs.c [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+#include "sw/device/lib/base/macros.h"
+#include "sw/device/lib/runtime/hart.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/test_framework/FreeRTOSConfig.h"
+
+static uint32_t get_mtval(void) {
+  uint32_t mtval;
+  CSR_READ(CSR_REG_MTVAL, &mtval);
+  return mtval;
+}
```
Comment (sriyerg) [inline sw/device/lib/testing/test_framework/ottf_isrs.c:17]: This seems better placed in `ibex.c/h`?

File: sw/device/lib/testing/test_framework/ottf_isrs.h [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_TESTING_TEST_FRAMEWORK_OTTF_ISRS_H_
+#define OPENTITAN_SW_DEVICE_LIB_TESTING_TEST_FRAMEWORK_OTTF_ISRS_H_
+
+/**
+ * An OTTF exception type.
+ *
+ * This enum is used to decode RISC-V exception causes generated by Ibex.
+ */
+typedef enum ottf_exc_id {
```
Comment (sriyerg) [inline sw/device/lib/testing/test_framework/ottf_isrs.h:13]: How about we move this to `ibex.h`?

Comment (sriyerg) [reply sw/device/lib/testing/test_framework/ottf_isrs.h:13]: Actually nevermind - this is RISCV specifc, not Ibex specific.

**Review by sriyerg — APPROVED** [review]
Left 2 questions, otherwise LGTM, thanks!

File: sw/device/lib/testing/test_framework/ottf_isrs.c (hunk 9 above)
Comment (timothytrippel) [reply sw/device/lib/testing/test_framework/ottf_isrs.c]: Fixed.

File: sw/device/lib/testing/test_framework/ottf_isrs.c (hunk 8 above)
Comment (timothytrippel) [reply sw/device/lib/testing/test_framework/ottf_isrs.c]: You're right. Fixed.

**Review by timothytrippel — COMMENTED** [review]
> Left 2 questions, otherwise LGTM, thanks!

Fixed both.

**Comment by timothytrippel** [comment]
PR dependencies removed. Free to merge, once CI checks pass.

**Pull request merged**